crates/fluvio-spu/src/services/internal/update_consumer_offset_handler.rs

```diff
@@ -0,0 +1,71 @@
+use std::{io::Error as IoError, time::Duration};
+
+use fluvio_controlplane::CONSUMER_STORAGE_TOPIC;
+use fluvio_protocol::{
+    api::{RequestMessage, ResponseMessage},
+    link::ErrorCode,
+    record::ReplicaKey,
+    record::Offset,
+};
+use fluvio_storage::FileReplica;
+use fluvio_types::PartitionId;
+use tracing::{instrument, trace};
+
+use crate::{
+    core::DefaultSharedGlobalContext,
+    replication::leader::LeaderReplicaState,
+    kv::consumer::{ConsumerOffset, ConsumerOffsetKey},
+};
+
+use super::update_consumer_offset_request::{UpdateConsumerOffsetRequest, UpdateConsumerOffsetResponse};
+
+#[instrument(skip(req_msg, ctx))]
+pub(crate) async fn handle_update_consumer_offset_request(
+    req_msg: RequestMessage<UpdateConsumerOffsetRequest>,
+    ctx: DefaultSharedGlobalContext,
+) -> Result<ResponseMessage<UpdateConsumerOffsetResponse>, IoError> {
+    let UpdateConsumerOffsetRequest {
+        consumer_id,
+        offset,
+        ttl,
+        replica_id,
+    } = req_msg.request;
+
+    let consumers_replica_id =
+        ReplicaKey::new(CONSUMER_STORAGE_TOPIC, <PartitionId as Default>::default());
+
+    let error_code = if let Some(ref replica) = ctx.leaders_state().get(&consumers_replica_id).await
```

Comment:
Shouldn't this get offset kv rather than trying to get leaders state?  Why bother creating abstraction if this is not being used?

Reply:
In this way we guarantee, that offset storage is used only on SPU that is a replica leader.